File: market/tests/UnixGlobUtils.test.ts
import expect from "expect.js";
import { toRegexPattern } from "../utils/UnixGlobUtils";

const matchGlob = (path: string, globPatter: string): boolean => {
  return Boolean(path.match(toRegexPattern(globPatter)));
};

const isWrongPattern = (globPatter: string) => {
  try {
    return !new RegExp(toRegexPattern(globPatter));
  } catch (ignore) {
    return true;
  }
};

describe("Glob utils test", () => {
  it("base", () => {
    expect(matchGlob("foo.html", "foo.html")).to.eql(true);
    expect(matchGlob("foo.html", "foo.htm")).to.eql(false);
    expect(matchGlob("foo.html", "bar.html")).to.eql(false);
  });

  it("match zero or more characters", () => {
    expect(matchGlob("foo.html", "f*")).to.eql(true);
    expect(matchGlob("foo.html", "*.html")).to.eql(true);
    expect(matchGlob("foo.html", "foo.html*")).to.eql(true);
    expect(matchGlob("foo.html", "*foo.html")).to.eql(true);
    expect(matchGlob("foo.html", "*foo.html*")).to.eql(true);
    expect(matchGlob("foo.html", "*.htm")).to.eql(false);
    expect(matchGlob("foo.html", "f.*")).to.eql(false);
  });

  it("match one character", () => {
    expect(matchGlob("foo.html", "?oo.html")).to.eql(true);
    expect(matchGlob("foo.html", "??o.html")).to.eql(true);
    expect(matchGlob("foo.html", "???.html")).to.eql(true);
    expect(matchGlob("foo.html", "???.htm?")).to.eql(true);
    expect(matchGlob("foo.html", "foo.???")).to.eql(false);
  });

  it("group of subpatterns", () => {
    expect(matchGlob("foo.html", "foo{.html,.class}")).to.eql(true);
    expect(matchGlob("foo.html", "foo.{class,html}")).to.eql(true);
    expect(matchGlob("foo.html", "foo{.htm,.class}")).to.eql(false);
  });

  it("bracket expressions", () => {
    expect(matchGlob("foo.html", "[f]oo.html")).to.eql(true);
    expect(matchGlob("foo.html", "[e-g]oo.html")).to.eql(true);
    expect(matchGlob("foo.html", "[abcde-g]oo.html")).to.eql(true);
    expect(matchGlob("foo.html", "[abcdefx-z]oo.html")).to.eql(true);
    expect(matchGlob("foo.html", "[!a]oo.html")).to.eql(true);
    expect(matchGlob("foo.html", "[!a-e]oo.html")).to.eql(true);
    expect(matchGlob("foo-bar", "foo[-a-z]bar")).to.eql(true);
    expect(matchGlob("foo.html", "foo[!-]html")).to.eql(true);
  });

  it("groups of subpattern with bracket expressions", () => {
    expect(matchGlob("foo.html", "[f]oo.{[h]tml,class}")).to.eql(true);
    expect(matchGlob("foo.html", "foo.{[a-z]tml,class}")).to.eql(true);
    expect(matchGlob("foo.html", "foo.{[!a-e]tml,.class}")).to.eql(true);
  });

  it("assume special characters are allowed in file names", () => {
    expect(matchGlob("{foo}.html", "\\{foo*")).to.eql(true);
    expect(matchGlob("{foo}.html", "*\\}.html")).to.eql(true);
    expect(matchGlob("[foo].html", "\\[foo*")).to.eql(true);
    expect(matchGlob("[foo].html", "*\\].html")).to.eql(true);
  });

  it("errors", () => {
    expect(isWrongPattern("*[a--z]")).to.eql(true);
    expect(isWrongPattern("*[a--]")).to.eql(true);
    expect(isWrongPattern("*[a-z")).to.eql(true);
    expect(isWrongPattern("*{class,java")).to.eql(true);
    expect(isWrongPattern("*.{class,{.java}}")).to.eql(true);
    expect(isWrongPattern("*.html\\")).to.eql(true);
  });

  it("unix specific", () => {
    expect(matchGlob("/tmp/foo", "/tmp/*")).to.eql(true);
    expect(matchGlob("/tmp/foo/bar", "/tmp/**")).to.eql(true);
    expect(matchGlob("myfile?", "myfile\\?")).to.eql(true);
    expect(matchGlob("one\\two", "one\\\\two")).to.eql(true);
    expect(matchGlob("one*two", "one\\*two")).to.eql(true);
  });
});

File: market/test/java/ru/yandex/market/tsum/utils/UnixGlobUtilsTest.java
package ru.yandex.market.tsum.utils;

import java.util.regex.Pattern;

import org.junit.Assert;
import org.junit.Test;

/**
 * @author Mishunin Andrei <a href="mailto:mishunin@yandex-team.ru"></a>
 * @date 15.07.2019
 */
public class UnixGlobUtilsTest {

    @Test
    public void toRegexPatternTest() {
        // basic
        Assert.assertTrue(matchGlob("foo.html", "foo.html"));
        Assert.assertFalse(matchGlob("foo.html", "foo.htm"));
        Assert.assertFalse(matchGlob("foo.html", "bar.html"));

        // match zero or more characters
        Assert.assertTrue(matchGlob("foo.html", "f*"));
        Assert.assertTrue(matchGlob("foo.html", "*.html"));
        Assert.assertTrue(matchGlob("foo.html", "foo.html*"));
        Assert.assertTrue(matchGlob("foo.html", "*foo.html"));
        Assert.assertTrue(matchGlob("foo.html", "*foo.html*"));
        Assert.assertFalse(matchGlob("foo.html", "*.htm"));
        Assert.assertFalse(matchGlob("foo.html", "f.*"));

        // match one character
        Assert.assertTrue(matchGlob("foo.html", "?oo.html"));
        Assert.assertTrue(matchGlob("foo.html", "??o.html"));
        Assert.assertTrue(matchGlob("foo.html", "???.html"));
        Assert.assertTrue(matchGlob("foo.html", "???.htm?"));
        Assert.assertFalse(matchGlob("foo.html", "foo.???"));

        // group of subpatterns
        Assert.assertTrue(matchGlob("foo.html", "foo{.html,.class}"));
        Assert.assertTrue(matchGlob("foo.html", "foo.{class,html}"));
        Assert.assertFalse(matchGlob("foo.html", "foo{.htm,.class}"));

        // bracket expressions
        Assert.assertTrue(matchGlob("foo.html", "[f]oo.html"));
        Assert.assertTrue(matchGlob("foo.html", "[e-g]oo.html"));
        Assert.assertTrue(matchGlob("foo.html", "[abcde-g]oo.html"));
        Assert.assertTrue(matchGlob("foo.html", "[abcdefx-z]oo.html"));
        Assert.assertTrue(matchGlob("foo.html", "[!a]oo.html"));
        Assert.assertTrue(matchGlob("foo.html", "[!a-e]oo.html"));
        Assert.assertTrue(matchGlob("foo-bar", "foo[-a-z]bar"));
        Assert.assertTrue(matchGlob("foo.html", "foo[!-]html"));

        // groups of subpattern with bracket expressions
        Assert.assertTrue(matchGlob("foo.html", "[f]oo.{[h]tml,class}"));
        Assert.assertTrue(matchGlob("foo.html", "foo.{[a-z]tml,class}"));
        Assert.assertTrue(matchGlob("foo.html", "foo.{[!a-e]tml,.class}"));

        // assume special characters are allowed in file names
        Assert.assertTrue(matchGlob("{foo}.html", "\\{foo*"));
        Assert.assertTrue(matchGlob("{foo}.html", "*\\}.html"));
        Assert.assertTrue(matchGlob("[foo].html", "\\[foo*"));
        Assert.assertTrue(matchGlob("[foo].html", "*\\].html"));

        // errors
        Assert.assertTrue(isWrongPattern("*[a--z]"));
        Assert.assertTrue(isWrongPattern("*[a--]"));
        Assert.assertTrue(isWrongPattern("*[a-z"));
        Assert.assertTrue(isWrongPattern("*{class,java"));
        Assert.assertTrue(isWrongPattern("*.{class,{.java}}"));
        Assert.assertTrue(isWrongPattern("*.html\\"));

        // unix specific
        Assert.assertTrue(matchGlob("/tmp/foo", "/tmp/*"));
        Assert.assertTrue(matchGlob("/tmp/foo/bar", "/tmp/**"));
        Assert.assertTrue(matchGlob("myfile?", "myfile\\?"));
        Assert.assertTrue(matchGlob("one\\two", "one\\\\two"));
        Assert.assertTrue(matchGlob("one*two", "one\\*two"));
    }

    private boolean matchGlob(String path, String globPatter) {
        String regexpPattern = UnixGlobUtils.toRegexPattern(globPatter);
        return Pattern.compile(regexpPattern).matcher(path).find();
    }

    private boolean isWrongPattern(String globPatter) {
        try {
            String regexpPattern = UnixGlobUtils.toRegexPattern(globPatter);
            Pattern.compile(regexpPattern);
            return false;
        } catch (Exception ignore) {
        }
        return true;
    }

}
